src/queries/role.ts

```diff
@@ -28,22 +29,76 @@ const matchAccessibleNameIfNeeded = (
   );
 };
 
+const accessibilityStates = [
+  'disabled',
+  'selected',
+  'checked',
+  'busy',
+  'expanded',
+] as const;
+
 const queryAllByRole = (
   instance: ReactTestInstance
 ): ((role: TextMatch, options?: ByRoleOptions) => Array<ReactTestInstance>) =>
   function queryAllByRoleFn(role, options) {
-    return instance.findAll(
-      (node) =>
+    return instance.findAll((node) => {
+      // run the cheapest checks first, and early exit too avoid unneeded computations
+      const matchRole =
         typeof node.type === 'string' &&
-        matchStringProp(node.props.accessibilityRole, role) &&
-        matchAccessibleNameIfNeeded(node, options?.name)
-    );
+        matchStringProp(node.props.accessibilityRole, role);
+
+      if (!matchRole) return false;
+
+      const mismatchAccessibilityState = accessibilityStates.some(
+        (accessibilityState) => {
+          const queriedState = options?.[accessibilityState];
+
+          if (typeof queriedState !== 'undefined') {
+            return (
+              queriedState !==
+              // default to false because disabled:undefined is equivalent to disabled:false
+              (node.props.accessibilityState?.[accessibilityState] ?? false)
```

Comment:
Seems like `checked` does not have default to `false` behaviour in real world tests for both iOS and Android. See more details [here](https://github.com/callstack/react-native-testing-library/wiki/Accessibility:-State). Doing some quick testing on iOS only, also `expanded` button does not default to `false`. For `busy`, `disabled` and `selected` it does however default to `false`.

I wonder what to do with that fact. We might want to default to `false` only for `busy`, `disabled` and `selected`